src/cargo/core/resolver/activation_key.rs

```diff
@@ -0,0 +1,73 @@
+use std::collections::HashSet;
+use std::hash::{Hash, Hasher};
+use std::num::NonZeroU64;
+use std::sync::{Mutex, OnceLock};
+
+use crate::core::SourceId;
+use crate::util::interning::InternedString;
+
+static ACTIVATION_KEY_CACHE: OnceLock<Mutex<HashSet<&'static ActivationKeyInner>>> =
+    OnceLock::new();
+
+type ActivationKeyInner = (InternedString, SourceId, SemverCompatibility);
+
+/// The activated version of a crate is based on the name, source, and semver compatibility.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ActivationKey {
+    inner: &'static ActivationKeyInner,
+}
+
+impl From<ActivationKeyInner> for ActivationKey {
+    fn from(inner: ActivationKeyInner) -> Self {
+        let mut cache = ACTIVATION_KEY_CACHE
+            .get_or_init(|| Default::default())
+            .lock()
+            .unwrap();
+        let inner = cache.get(&inner).cloned().unwrap_or_else(|| {
+            let inner = Box::leak(Box::new(inner));
+            cache.insert(inner);
+            inner
+        });
+        Self { inner }
+    }
+}
+
+/// A type that represents when cargo treats two versions as compatible.
+/// Versions `a` and `b` are compatible if their left-most nonzero digit is the same.
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug, PartialOrd, Ord)]
+pub enum SemverCompatibility {
+    Major(NonZeroU64),
+    Minor(NonZeroU64),
+    Patch(u64),
+}
+
+impl From<&semver::Version> for SemverCompatibility {
+    fn from(ver: &semver::Version) -> Self {
+        if let Some(m) = NonZeroU64::new(ver.major) {
+            return SemverCompatibility::Major(m);
+        }
+        if let Some(m) = NonZeroU64::new(ver.minor) {
+            return SemverCompatibility::Minor(m);
+        }
+        SemverCompatibility::Patch(ver.patch)
+    }
+}
+
+impl ActivationKey {
+    /// Returns a value that implements a "no hash" hashable value.
+    /// This is possible since all `ActivationKey` are already interned in a `HashSet`.
+    pub fn no_hash(self) -> ActivationKeyNoHash {
+        ActivationKeyNoHash(std::ptr::from_ref(self.inner) as u64)
+    }
+}
+
+#[derive(Clone, Copy, Eq, PartialEq)]
+pub struct ActivationKeyNoHash(u64);
```

Comment:
Should this be a separate type? Or can it just be the hash implementation for `ActivationKey`?